### PR TITLE
WIP: Added Patch Manager

### PR DESCRIPTION
**Pull request by timschumi**
DO NOT MERGE!
Still needs cleanup
I added it into the Build Menu for now

### CODE REVIEW

**Comment by timschumi** [comment]
8e806f2: It's not required to set the device before accessing the PAtch Manager anymore. Also it now searches in every device tree and in the patches folder in the root directory

Also, I wanted to add a check at exit-time which checks for still applied patches, but I'm not sure how I should access the necessary functions

**Comment by a7r3** [comment]
How about creation of patches on Successful Build completion ?

**Comment by timschumi** [comment]
How should ScriBt know
1) If the build is ok?
2) What needs to get patched?

Or do you mean that the compiled ROM should be patched?

**Comment by a7r3** [comment]
1. The [exit code](https://github.com/a7r3/ScriBt/blob/master/ROM.sh#L707) given by make

`if(exit code == 0) then success else fail`

2. Changes made to the source ?
Once the build is complete, a git diff is taken on a set* of repos, and a set of corresponding patches are created

\* It can be a specific set of repos, strategy required to determine such repos

**Comment by akhilnarang** [comment]
repo forall -c "git clean -df && git reset --hard HEAD"
That'll take a while though, as it'll work on all ~500 repos

**Comment by akhilnarang** [comment]
Or if you get the path of the repos where the patches are applied, can just run a for loop across those and reset them

**Comment by a7r3** [comment]
Patches are placed inside the directory which needs to get patched, so determining the location is done here itself

**Comment by timschumi** [comment]
@a7r3
Exit Code == 0 doesn't mean the build will work on such devices. But that is basically a good idea. I'll think about it further

@akhilnarang
The first will take some time indeed and the second one was my original Idea. The functions of the patchmgr should still be defined (as they were never unset) so I could still Access them. Will think about this, too.

**Comment by akhilnarang** [comment]
@TimSchumi 
exit code == 0 will just mean make ran successfully, whether the build is actually fine or not you'll need to flash to find out x'D

**Comment by timschumi** [comment]
Yeah, that's what I meant with the question How ScriBt could determine if the build is ok?

**Comment by a7r3** [comment]
Bootability isn't in ScriBt's hands, we should get that input from the user

**Comment by timschumi** [comment]
That's what I'm planning

**Comment by a7r3** [comment]
Bootability by asking for Input ? Alright then
Can be asked when the user executes ScriBt, goes to Patch Manager, and Patch Manager detects the presence of a "Save File" or something which...

* Indicates that user has built a ROM with some changes and has ordered to save 'em changes when the user says that "It Booted"
* Has the patch data (Changes done in the entire source) itself

**Comment by timschumi** [comment]
First I'll do the Patch reversing, then I'll do the Patch Creator

**Comment by timschumi** [comment]
I'm done with the new additions for now. I would appreciate if you could review them before merging those :)

**Comment by a7r3** [comment]
Sure :+1:
I'd call @akhilnarang too

**Comment by timschumi** [comment]
I'll search for a better way to get the repository names though

"repo forall" seems very slow and inefficient

Unfortunately I can't do anything about files outside repositories

**Comment by akhilnarang** [comment]
repo forall -j${THREADS}

THREADS can be given the value of nproc, or grep -c "^processor"
/proc/cpuinfo.
On 5 Mar 2017 22:22, "Tim Schumacher" <notifications@github.com> wrote:

I'll search for a better way to get the repository names though

"repo forall" seems very slow and inefficient

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/a7r3/ScriBt/pull/25#issuecomment-284242776>, or mute
the thread

**Comment by timschumi** [comment]
Thanks, but with slow and inefficient I meant that it is very slow when the computer is also running some other stuff in the background (especially things that cause heavy load on the HDD). A single command (which doesn't spawn many subshells and multiple heavy process calls unlike forall) would be much more more fast, when forall with multiple threads would slow down even more

If no other tasks are running, a single-threaded forall command is fast too

EDIT: Found it, it was "repo list -p"

**Comment by a7r3** [comment]
@TimSchumi /me Newbie to **GitHub reviews**, bear with it xD
Please perform the changes mentioned in the review, I'd test-run the patch creation command soon, Syncing sources ATM.

Rest of the patch manager is Alright for me, need to understand `patch_creator()` before reviewing it.

**Suggestion :**
- New line before and after function

```
# Code
<newline>
function itis()
{
    # code
} # itis
<newline>
# Code
```

**Comment by timschumi** [comment]
@a7r3
I'll add some comments to the code so you can understand it better :)

I don't know when though...

EDIT3: Done. MGit 4 the win!

EDIT: I didn't mean the GitHub reviews btw. I just wanted that you look over the changes and test them

EDIT2: Might I help a bit with PyScriBt?

**Comment by a7r3** [comment]
* Can you provide an Example of how would a Patch file look like ? (After Creation)

> Altered the [main command](https://github.com/a7r3/ScriBt/pull/25/files#diff-07c55c018514d9355fb188cfe2d6e244R969) for usability and got this as the Patch file

```
CALL_ME_ROOT=$(pwd);
while read PROJECT; do 
    cd ${CALL_ME_ROOT}/${PROJECT}; 
    git diff |
    sed -e "s@ a/@ a/${PROJECT}/@g" | 
    sed -e "s@ b/@ b/${PROJECT}/@g" >> patch_${COUNT};
    echo -en "\033[KGenerated patch for repo $COUNT of $PROJECT_COUNT\r";
    cat ${CALL_ME_ROOT}/${PROJECT}/patch_${COUNT};
    ((COUNT++));
done <<< "${PROJECTS}"
```

```
diff --git a/vendor/tesla/tesla.devices b/vendor/tesla/tesla.devices
index 3d53650..06fa54f 100644
--- a/vendor/tesla/tesla.devices
+++ b/vendor/tesla/tesla.devices
@@ -1,3 +1,10 @@
+kangler
+krexus
+kokay
+k
+kay
+kraft
+kernel
 angler
 bacon
 bullhead
```

> Just an Example xD

No plans for starting **pyScriBt** as of now, need some free time to _Python-ify_ the Bash Code. When that's done, you guys may make changes in it. Of course, there would be a lot of changes as I'm learning Python.

EDIT1: But, I'd mention you if I publish an Initial **pyScriBt**. Initially, It'd just be a migration from Bash to Python, Development starts right after that.

Thanks for showing your Interest :smile: :+1:

File: ROM.sh [hunk 1]
```diff
@@ -886,6 +893,113 @@ function build() # 4
...
+                    PROJECTS="$(repo list -p)"; # Get all teh projects
+                    PROJECT_COUNT=$(wc -l <<< "$PROJECTS"); # Count all teh projects
+                    [ -f "${CALL_ME_ROOT}/${PATCH_PATH}" ] && rm -rf ${CALL_ME_ROOT}/${PATCH_PATH} # Delete existing patch
+                    COUNT=1
+                    echo "";
+                    while read PROJECT; do # Foreach does not work, as it seems to spawn a subshell
+                        cd ${CALL_ME_ROOT}/${PROJECT}
+                        git diff | 
+                          sed -e "s@ a/@ a/${PROJECT}/@g" |
+                          sed -e "s@ b/@ b/${PROJECT}/@g" >> ${CALL_ME_ROOT}/${PATCH_PATH}; # Extend a/ and b/ with the project's path, as git diff only outputs the paths relative to the git repository's root
```
Comment (a7r3) [inline ROM.sh:971]: Understood :+1:

**Comment by timschumi** [comment]
The patch file looks good. (Is vendor/tesla a git repo?)
Could you explain to me what and why you changed things in the method?

About pyScriBt:
I think I'll start a bit with rewriting it from scratch. I guess transferring stuff from bash won't work really well.
Would a ScriBt GUI too overpowered?

**Comment by a7r3** [comment]
That's just for usability, since I **only** executed that snippet

* Set `CALL_ME_ROOT` to Working Dir
* `PATCH_PATH` is not set, so I set the destination to `patch_${COUNT}` instead
* `cat` command was for checking the output. Test-run purposes.
* Manually set `PROJECTS=$(repo list -p)`

The Code is Logically the same, unaffected. Changed the Patch Destination file only.

**Comment by timschumi** [comment]
Oh, then it's ok. I was just wondering...
I'll close the Feature Discussion Issue now

**Comment by a7r3** [comment]
Alright, time for DerpCheck, and I'd execute this version of ScriBt.

**Comment by timschumi** [comment]
Before you merge it: What would you think about changing it to all subdir git repositories instead of just those mentioned in the manifest?

**Comment by a7r3** [comment]
In reply to Second Last comment...

* Yes, [vendor/tesla](https://github.com/TeslaROM-N/vendor_tesla) is a Git Repo.

* You can start with pyScriBt, no issues.

I "think" a GUI would be, uhmm,.... Let's think on it.
Shall we conduct a poll ? xD :thinking:

**Comment by timschumi** [comment]
Nah, let's finish cmdline pyScriBt first. The GUI can still be built on top

This should be ready to merge (I'll not do the "search for git repositories" thing to prevent issues)

**Comment by a7r3** [comment]
That's alright, I'd check-merge it tonight.
Good Morning

**Comment by a7r3** [comment]
Merging it :v: 
Documentation in Progress

**Pull request merged**